### PR TITLE
Remove unsound offset_of macro

### LINKED ISSUE #1483: offset_of is unsound
The macro at https://github.com/tokio-rs/mio/blob/1667a7027382bd43470bc43e5982531a2e14b7ba/src/sys/windows/named_pipe.rs#L26 is unsound and causes UB due to using `*` on a NULL pointer. See [the reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html) for details.

The [memoffset crate](https://github.com/Gilnaa/memoffset/) provides a version of this macro that avoids UB (on older versions of rustc, this is not always possible, so it falls back to the 'least incorrect' version when needed).

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
And replace it with constants that define the offsets to the fields.
It's not a pretty solution, but it's one without UB.

Closes #1483.

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
This should fix #1483 @RalfJung.

File: src/sys/windows/named_pipe.rs [hunk 1]
```diff
@@ -645,7 +649,7 @@ fn write_done(status: &OVERLAPPED_ENTRY, events: Option<&mut Vec<Event>>) {
     // Acquire the `Arc<Inner>`. Note that we should be guaranteed that
     // the refcount is available to us due to the `mem::forget` in
     // `schedule_write` above.
-    let me = unsafe { overlapped2arc!(status.overlapped(), Inner, write) };
+    let me = unsafe { Arc::from_raw((status.overlapped() as usize - WRITE_OFFSET) as *mut Inner) };
```
Comment (Darksonn) [inline src/sys/windows/named_pipe.rs:652]: I would prefer going through `*mut u8` here. Integer to pointer casts are sketchy soundness-wise.
```suggestion
    let me = unsafe { Arc::from_raw(status.overlapped().cast::<u8>().sub(WRITE_OFFSET).cast::<Inner>()) };
```
or if the constants are 0, 1, 2, you can even do this.
```suggestion
    let me = unsafe { Arc::from_raw(status.overlapped().sub(WRITE_OFFSET).cast::<Inner>()) };
```

Comment (RalfJung) [reply src/sys/windows/named_pipe.rs:652]: Any reason not to use `wrapping_sub` which is less subtle to call?

Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs:652]: I agree the code is somewhat sketchy, but that really the nature of this code. Going via `sub` is a mistake as that uses `T` * count. Even if you cast to a `u8`, like you suggested, it's too easy to mess up imo.

File: src/sys/windows/named_pipe.rs [hunk 2]
```diff
@@ -83,21 +64,43 @@ pub struct NamedPipe {
...
     io: Mutex<Io>,
-
     pool: Mutex<BufferPool>,
 }
 
+/// Offsets from a pointer to `Inner` to the `connect`, `read` and `write`
+/// fields.
+const CONNECT_OFFSET: usize = 0;
+const READ_OFFSET: usize = CONNECT_OFFSET + size_of::<Overlapped>(); // `connect` fields.
+const WRITE_OFFSET: usize = READ_OFFSET + size_of::<Overlapped>(); // `read` fields.
```
Comment (RalfJung) [inline src/sys/windows/named_pipe.rs:86]: This seems rather fragile...

If I understand correctly, we actually do have an instance of the relevant type available at the place where the offset needs to be computed. So why can't the code simply compute the offsets based on that instance? `offset_of!` is only hard when you don't have a value of the type you want to compute offsets in.

Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs:86]: I agree it's fragile.

But we don't have an instance available in [`read_done`](https://github.com/tokio-rs/mio/pull/1485/files#diff-8648c6b3c0e6fc7faf5a29bd380248e904852cb3bfaf6e0e2638af9d281b4960R614) when we need it. We have a pointer to one of the fields of the instance. How do you suggest we get the base pointer (to `Inner`) then?

File: src/sys/windows/named_pipe.rs (hunk 1 above)
Comment (Darksonn) [reply src/sys/windows/named_pipe.rs:652]: You could define helper functions for wrapping the subtraction and test them in the test :woman_shrugging:

File: src/sys/windows/named_pipe.rs (hunk 2 above)
Comment (RalfJung) [reply src/sys/windows/named_pipe.rs:86]: Ah, I thought you'd have an instance anywhere, sorry... I don't really understand the code.^^

Btw, this looks like `container_of!`-style pointer arithmetic, which does not work well with the Rust aliasing model... https://github.com/rust-lang/unsafe-code-guidelines/issues/243.

But ignoring aliasing issues, if you do not want to depend on `addr_of!`, you could either decide to not care about code being technically UB (you could still avoid the NULL ptr by using `NonNull::dangling` as the base) -- or I guess hard-code the offsets as you do...

File: src/sys/windows/named_pipe.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs]: @Darksonn I've replaced the constants with methods and updated the test above.

File: src/sys/windows/named_pipe.rs (hunk 2 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs]: I don't really know of a way to solve this. I'm following https://github.com/rust-lang/unsafe-code-guidelines/issues/243 now so if someone suggests a solution that is not UB we can always update this code.

File: src/sys/windows/named_pipe.rs [hunk 3]
```diff
@@ -83,21 +64,72 @@ pub struct NamedPipe {
...
+    /// Caller must ensure `ptr` is pointing to `Inner.connect`.
+    unsafe fn ptr_from_conn_overlapped(ptr: *mut OVERLAPPED) -> *const Inner {
+        // `connect` is the first field, so the pointer are the same.
+        ptr.cast()
+    }
+
+    /// Same as [`ptr_from_conn_overlapped`] but for `Inner.read`.
+    unsafe fn ptr_from_read_overlapped(ptr: *mut OVERLAPPED) -> *const Inner {
+        // `read` is after `connect: Overlapped`.
+        (ptr as usize + size_of::<Overlapped>()) as *const Inner
```
Comment (Darksonn) [inline src/sys/windows/named_pipe.rs:99]: How about this? It avoids the cast through `usize`.
```suggestion
        (ptr as *const Overlapped).wrapping_sub(1) as *const Inner
```

Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs]: Done.

**Comment by Thomasdezeeuw** [comment]
Merging #1487 because that CI fails without this change, and this pr fails without #1487.

File: src/sys/windows/named_pipe.rs [hunk 4]
```diff
@@ -83,21 +63,72 @@ pub struct NamedPipe {
     inner: Arc<Inner>,
 }
 
+/// # Notes
+///
+/// The memory layout of this structure must be fixed as the `*_OFFSET`
+/// constants depend on it, see the `offset_constants` test.
 #[repr(C)]
 struct Inner {
```
Comment (Darksonn) [inline src/sys/windows/named_pipe.rs:71]: Isn't this comment out of date?

**Review by Darksonn — APPROVED** [review]

File: src/sys/windows/named_pipe.rs (hunk 4 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/named_pipe.rs]: Good catch, updated.

**Pull request merged**